CLDSRV-362 - Throw error when copying a cold or transitioning object

welansari opened the pull request:
A cold object should not be allowed to get copied as the data is not accessible.

Same issue happens when copying an object that is transitioning, the data might get deleted while copying is still in progress.

Issue: CLDSRV-362

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-362 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.7.10`


Please check the `Fix Version/s` of CLDSRV-362, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-362 contains:


* `8.7.10`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.7.11`


Please check the `Fix Version/s` of CLDSRV-362, or the target
branch of this pull request.

francoisferrand commented on lib/api/objectCopy.js:
```diff
@@ -290,6 +292,18 @@ function objectCopy(authInfo, request, sourceBucket,
                         return next(err, null, destBucketMD);
                     }
+                    // return error when object is cold
+                    if (locationConstraints[sourceObjMD.dataStoreName]?.isCold) {
```
that check may soon be incorrect...

at the moment, indeed the `dataStoreName` is updated when the object is restored (so it will not 'appear' as cold), but this causes other issues : I think there is WIP to avoid changing the data store when restoring an object, so that we keep track of underlying cold location anyway.... and that check would thus fail for cold location.

Not sure when that is expected though, maybe @nicolas2bert can confirm if this will conflict and/or if there is a better to check for "object is cold AND not restored"

francoisferrand commented on tests/functional/aws-node-sdk/test/object/objectCopy.js:
```diff
@@ -1234,6 +1235,42 @@ describe('Object Copy', () => {
+                    });
+            });
+        });
+
```
need another test, to ensure we can indeed copy a "cold & restored" object
(may be the same check for the moment, but that may not always be true...)

francoisferrand commented on tests/functional/aws-node-sdk/test/object/copyPart.js:
```diff
@@ -710,6 +711,72 @@ describe('Object Part Copy', () => {
+            });
+        });
+
+        it('should copy restored object', done => {
```
nit: "should copy part from restored object"

francoisferrand commented on lib/api/objectPutCopyPart.js:
```diff
@@ -133,6 +133,22 @@ function objectPutCopyPart(authInfo, request, sourceBucket,
                         sourceLocationConstraintName =
                             sourceObjMD.location[0].dataStoreName;
                     }
+                    // return error when object is cold
+                    if (sourceObjMD.archive &&
+                        // Object is in cold backend
+                        (!sourceObjMD.archive.restoreRequestedAt ||
+                        // Object is being restored
+                        (sourceObjMD.archive.restoreRequestedAt && !sourceObjMD.archive.restoreCompletedAt))) {
```
should we introduce a function for this? (e.g. `isObjecDatatInColdStorage`)

or possibly a helper which verifies both that data is present and object is not being transitioned and returns the appropriate error? (e.g. `VerifyColdObjectAvailable`) ?

francoisferrand commented on lib/api/objectPutCopyPart.js:
```diff
@@ -133,6 +133,22 @@ function objectPutCopyPart(authInfo, request, sourceBucket,
+                    // return error when object is transitioning to a second location
+                    if (sourceObjMD['x-amz-scal-transition-in-progress']) {
+                        const err = errors.InvalidObjectState
+                        .customizeDescription('The operation is not valid on a transitioning object');
```
nit: missing indent

Commit:
add helper to check if object is in cold storage

Issue: CLDSRV-362

Commit:
throw error when copying a cold or transitioning object

A cold object should not be allowed to get copied as the data
is not accessible.

Same issue happens when copying an object that is transitioning,
the data might get deleted while copying is still in progress.

Issue: CLDSRV-362

Commit:
throw error when copying parts from a cold object

A cold object should not be allowed to get copied as the data
is not accessible.

Issue: CLDSRV-362

Commit:
throw error when getting a transitioning object

Issue: CLDSRV-362

francoisferrand commented on lib/api/objectPutCopyPart.js:
```diff
@@ -133,6 +134,11 @@ function objectPutCopyPart(authInfo, request, sourceBucket,
+                    // check if object data is in a cold storage
+                    const coldErr = verifyColdObjectAvailable(sourceObjMD);
+                    if (coldErr) {
+                        return callback(coldErr, null);
```
shouldn't it be `next(coldErr, null)` ?

welansari replied:
it should

francoisferrand commented on tests/functional/aws-node-sdk/test/object/copyPart.js:
```diff
@@ -710,6 +711,72 @@ describe('Object Part Copy', () => {
                 });
             });
 
+
```
nit: extra empty line

francoisferrand commented on tests/unit/api/objectGet.js:
```diff
@@ -423,6 +423,24 @@ describe('objectGet API', () => {
+            });
+        });
+    });
+
```
should test also the other cases? (cold, restored)

welansari replied:
test for those cases already exist

francoisferrand commented on tests/functional/aws-node-sdk/test/object/copyPart.js:
```diff
@@ -710,6 +711,72 @@ describe('Object Part Copy', () => {
             });
 
+
+        it('should not copy a cold object', done => {
```
copy part of

francoisferrand commented on tests/functional/aws-node-sdk/test/object/copyPart.js:
```diff
@@ -710,6 +711,72 @@ describe('Object Part Copy', () => {
+            });
+        });
+
+        it('should not copy an object when it\'s transitioning to cold', done => {
```
copy part of

francoisferrand reviewed (approved):
LGTM

KazToozs commented on lib/api/apiUtils/object/coldStorage.js:
```diff
@@ -198,9 +197,34 @@ function startRestore(objectMD, restoreParam, log, cb) {
+        // Object is in cold backend
+        (!objMD.archive.restoreRequestedAt ||
+        // Object is being restored
+        (objMD.archive.restoreRequestedAt && !objMD.archive.restoreCompletedAt))) {
```
For my understanding, why is the existence of `objMD.archive` alone not enough to throw error? Is it maintained as an attribute on the object once a restore is successful?

welansari replied:
When an object is restored `objMD.archive` is still present in the metadata, and points to the data on the cold side.
Restoration is a temporary state that eventually expires, that's why we need to keep it.

KazToozs commented on lib/api/objectCopy.js:
```diff
@@ -290,6 +291,11 @@ function objectCopy(authInfo, request, sourceBucket,
                     }
+                    // check if object data is in a cold storage
+                    const coldErr = verifyColdObjectAvailable(sourceObjMD);
+                    if (coldErr) {
```
Nit: is it necessary/good practice to create a variable? Ignore if so
```suggestion
                    if (verifyColdObjectAvailable(sourceObjMD)) {
```

KazToozs replied:
same applies for some areas below

welansari replied:
the function returns an error so we have to store its value somewhere to return it

KazToozs reviewed (commented):
LGTM otherwise

Commit:
bump version to 8.7.11

Issue: CLDSRV-362

welansari commented:
/approve

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-362 contains:


* `8.7.10`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.7.11`


Please check the `Fix Version/s` of CLDSRV-362, or the target
branch of this pull request.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.7`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.4`
* `development/8.5`
* `development/8.6`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.7`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.4`
* `development/8.5`
* `development/8.6`



Please check the status of the associated issue CLDSRV-362.

Goodbye kerkesni.
